File: golang/llamacpp/LlamaCppClient.go
package llamacpp

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
)

const (
	CompletionDataPrefix = "data: "
)

var (
	ErrorNon200Response = errors.New("Non-200 response from llama.cpp")
)

type LlamaCppClient struct {
	HttpClient            *http.Client
	LlamaCppConfiguration *LlamaCppConfiguration
}

func (self *LlamaCppClient) GetHealth(
	ctx context.Context,
	responseChannel chan<- LlamaCppHealthStatus,
) {
	request, err := http.NewRequestWithContext(
		ctx,
		"GET",
		self.LlamaCppConfiguration.HttpAddress.BuildUrlWithPath("health").String(),
		nil,
	)

	if err != nil {
		responseChannel <- LlamaCppHealthStatus{
			Error:        err,
			ErrorMessage: err.Error(),
			Status:       Error,
		}

		return
	}

	response, err := self.HttpClient.Do(request)

	if err != nil {
		responseChannel <- LlamaCppHealthStatus{
			Error:        err,
			ErrorMessage: err.Error(),
			Status:       Error,
		}

		return
	}

	defer response.Body.Close()

	if http.StatusOK != response.StatusCode {
		responseChannel <- LlamaCppHealthStatus{
			Error:        ErrorNon200Response,
			ErrorMessage: ErrorNon200Response.Error(),
			Status:       Error,
		}

		return
	}

	var llamaCppHealthStatus LlamaCppHealthStatus

	err = json.NewDecoder(response.Body).Decode(&llamaCppHealthStatus)

	if err != nil {
		responseChannel <- LlamaCppHealthStatus{
			Error:        err,
			ErrorMessage: err.Error(),
			Status:       Error,
		}

		return
	}

	responseChannel <- llamaCppHealthStatus
}

func (self *LlamaCppClient) GetSlots(
	ctx context.Context,
	responseChannel chan<- LlamaCppSlotStatus,
) {
	defer close(responseChannel)

	request, err := http.NewRequestWithContext(
		ctx,
		"GET",
		self.LlamaCppConfiguration.HttpAddress.BuildUrlWithPath("slots").String(),
		nil,
	)

	if err != nil {
		responseChannel <- LlamaCppSlotStatus{
			Error:        err,
			ErrorMessage: err.Error(),
			State:        0,
		}

		return
	}

	if self.LlamaCppConfiguration.ApiKey != "" {
		request.Header.Set("Authorization", "Bearer "+self.LlamaCppConfiguration.ApiKey)
	}

	response, err := self.HttpClient.Do(request)

	if err != nil {
		responseChannel <- LlamaCppSlotStatus{
			Error:        err,
			ErrorMessage: err.Error(),
			State:        0,
		}

		return
	}

	defer response.Body.Close()

	if http.StatusOK != response.StatusCode {
		responseChannel <- LlamaCppSlotStatus{
			Error:        ErrorNon200Response,
			ErrorMessage: ErrorNon200Response.Error(),
			State:        0,
		}

		return
	}

	var llamaCppSlotStatuses []LlamaCppSlotStatus

	err = json.NewDecoder(response.Body).Decode(&llamaCppSlotStatuses)

	if err != nil {
		responseChannel <- LlamaCppSlotStatus{
			Error:        err,
			ErrorMessage: err.Error(),
			State:        0,
		}

		return
	}

	for _, llamaCppSlotStatus := range llamaCppSlotStatuses {
		responseChannel <- llamaCppSlotStatus
	}
}

func (self *LlamaCppClient) GetSlotsAggregatedStatus(
	ctx context.Context,
	responseChannel chan<- LlamaCppSlotsAggregatedStatus,
) {
	var llamaCppSlotsAggregatedStatus LlamaCppSlotsAggregatedStatus

	slotsChannel := make(chan LlamaCppSlotStatus)

	go self.GetSlots(ctx, slotsChannel)

	for slotStatus := range slotsChannel {
		if slotStatus.Error != nil {
			responseChannel <- LlamaCppSlotsAggregatedStatus{
				Error:        slotStatus.Error,
				ErrorMessage: slotStatus.ErrorMessage,
				Status:       Error,
			}

			return
		}

		if slotStatus.IsProcessing() {
			llamaCppSlotsAggregatedStatus.SlotsProcessing += 1
		} else {
			llamaCppSlotsAggregatedStatus.SlotsIdle += 1
		}
	}

	responseChannel <- llamaCppSlotsAggregatedStatus
}

File: src/llamacpp/llamacpp_client.rs
use std::net::SocketAddr;
use std::time::Duration;

use anyhow::anyhow;
use anyhow::Result;
use reqwest::header;
use url::Url;

use crate::llamacpp::slot::Slot;
use crate::llamacpp::slots_response::SlotsResponse;
use crate::llamacpp::models_response::ModelsResponse;

pub struct LlamacppClient {
    client: reqwest::Client,
    slots_endpoint_url: String,
    models_endpoint_url: String,
}

impl LlamacppClient {
    pub fn new(addr: SocketAddr, api_key: Option<String>) -> Result<Self> {
        let mut builder = reqwest::Client::builder().timeout(Duration::from_secs(3));
        let mut headers = header::HeaderMap::new();

        headers.insert(
            header::ACCEPT,
            header::HeaderValue::from_static("application/json"),
        );

        if let Some(api_key_value) = api_key {
            let mut auth_value = header::HeaderValue::from_str(&format!("Bearer {api_key_value}"))?;

            auth_value.set_sensitive(true);

            headers.insert(header::AUTHORIZATION, auth_value);
        }

        builder = builder.default_headers(headers);

        Ok(Self {
            client: builder.build()?,
            slots_endpoint_url: Url::parse(&format!("http://{addr}/slots"))?.to_string(),
            models_endpoint_url: Url::parse(&format!("http://{addr}/v1/models"))?.to_string(),
        })
    }

    pub async fn get_available_slots(&self) -> SlotsResponse {
        let url = self.slots_endpoint_url.to_owned();

        let response = match self.client.get(url.clone()).send().await {
            Ok(resp) => resp,
            Err(err) => {
                return SlotsResponse {
                    error: Some(format!("Request to {url} failed: {err}")),
                    is_authorized: None,
                    is_connect_error: Some(err.is_connect()),
                    is_decode_error: Some(err.is_decode()),
                    is_deserialize_error: None,
                    is_request_error: Some(err.is_request()),
                    is_unexpected_response_status: None,
                    is_slot_endpoint_enabled: Some(true),
                    slots: vec![],
                };
            }
        };

        let status = response.status();

        match status {
            reqwest::StatusCode::OK => {
                let (slots, error) = match response.json::<Vec<Slot>>().await {
                    Ok(slots) => (Some(slots), None),
                    Err(err) => (None, Some(err.to_string())),
                };

                SlotsResponse {
                    error: error.clone(),
                    is_authorized: Some(true),
                    is_unexpected_response_status: Some(false),
                    is_connect_error: Some(false),
                    is_decode_error: Some(false),
                    is_deserialize_error: Some(error.is_some()),
                    is_request_error: Some(false),
                    is_slot_endpoint_enabled: Some(true),
                    slots: slots.unwrap_or_default(),
                }
            }
            reqwest::StatusCode::UNAUTHORIZED => SlotsResponse {
                error: Some("Unauthorized".into()),
                is_authorized: Some(false),
                is_unexpected_response_status: Some(false),
                is_connect_error: Some(false),
                is_decode_error: Some(false),
                is_deserialize_error: None,
                is_request_error: Some(false),
                is_slot_endpoint_enabled: None,
                slots: vec![],
            },
            reqwest::StatusCode::NOT_IMPLEMENTED => SlotsResponse {
                error: Some("Not implemented".into()),
                is_authorized: None,
                is_unexpected_response_status: Some(false),
                is_connect_error: Some(false),
                is_decode_error: Some(false),
                is_deserialize_error: None,
                is_request_error: Some(false),
                is_slot_endpoint_enabled: Some(false),
                slots: vec![],
            },
            _ => SlotsResponse {
                error: Some("Unexpected response status".into()),
                is_authorized: None,
                is_unexpected_response_status: Some(true),
                is_connect_error: Some(false),
                is_decode_error: Some(false),
                is_deserialize_error: None,
                is_request_error: Some(false),
                is_slot_endpoint_enabled: None,
                slots: vec![],
            },
        }
    }

    pub async fn get_model(&self) -> Result<Option<String>> {
        let url = self.models_endpoint_url.to_owned();

        let response = match self.client.get(url.clone()).send().await {
            Ok(resp) => resp,
            Err(err) => {
                return Err(anyhow!(
                    "Request to '{}' failed: '{}'; connect issue: {}; decode issue: {}; request issue: {}; status issue: {}; status: {:?}",
                    url,
                    err,
                    err.is_connect(),
                    err.is_decode(),
                    err.is_request(),
                    err.is_status(),
                    err.status()
                ));
            }
        };

        match response.status() {
            reqwest::StatusCode::OK => {
                let models_response: ModelsResponse = response.json().await?;
                if let Some(models) = models_response.models {
                    if models.is_empty() {
                        Ok(None)
                    } else {
                        Ok(models.first().and_then(|m| Some(m.model.clone())))
                    }
                } else {
                    Ok(None)
                }
            },
            _ => Err(anyhow!("Unexpected response status")),
        }
    }
}
